src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -354,104 +459,192 @@ cbuf_mark_relinquish_all(struct cbuf_comp_info *cci)
 		cbi = bin->c;
 		do {
-			CBUF_FLAG_REM(cbi->owner.m, CBUF_RELINQ);
-			cbi = FIRST_LIST(cbi, next, prev);
+			if (!cbi) break;
+			m = &cbi->owner;
+			do {
+				CBUF_FLAG_ATOMIC_REM(m->m, CBUF_RELINQ);
+				m  = FIRST_LIST(m, next, prev);
+			} while(m != &cbi->owner);
+			cbi   = FIRST_LIST(cbi, next, prev);
 		} while (cbi != bin->c);
 	}
 }
 
+static inline void
+cbuf_thread_block(struct cbuf_comp_info *cci, unsigned long request_size)
+{
+	struct blocked_thd bthd;
+
+	bthd.thd_id       = cos_get_thd_id();
+	bthd.request_size = request_size;
+	rdtscll(bthd.blk_start);
+	ADD_LIST(&cci->bthd_list, &bthd, next, prev);
+	cci->num_blocked_thds++;
+	cbuf_mark_relinquish_all(cci);
+	CBUF_RELEASE();
+	sched_block(cos_spd_id(), 0);
+}
+
+/* wake up all blocked threads whose request size smaller than or equal to available size */
+static void
+cbuf_thd_wake_up(struct cbuf_comp_info *cci, unsigned long sz)
+{
+	struct blocked_thd *bthd, *next;
+	unsigned long long cur, tot;
+
+	assert(cci->num_blocked_thds >= 0);
+	/* Cannot wake up thd when in shrink */
+	assert(cci->target_size >= cci->allocated_size);
+
+	if (cci->num_blocked_thds == 0) return;
+	bthd = cci->bthd_list.next;
+	while (bthd != &cci->bthd_list) {
+		next = FIRST_LIST(bthd, next, prev);
+		if (bthd->request_size <= sz) {
+			REM_LIST(bthd, next, prev);
+			cci->num_blocked_thds--;
+			rdtscll(cur);
+			tot = cur-bthd->blk_start;
+			cci->blk_tot += tot;
+			if (tot > cci->blk_max) cci->blk_max = tot;
+			sched_wakeup(cos_spd_id(), bthd->thd_id);
+		}
+		bthd = next;
+	}
+	if (cci->num_blocked_thds == 0) cbuf_unmark_relinquish_all(cci);
+}
+
+static void cbuf_shrink(struct cbuf_comp_info *cci, int diff);
+
 int
-cbuf_create(spdid_t spdid, int size, long cbid)
+cbuf_create(spdid_t spdid, unsigned long size, int cbid)
 {
 	struct cbuf_comp_info *cci;
 	struct cbuf_info *cbi;
 	struct cbuf_meta *meta;
+	struct cbuf_bin *bin;
 	int ret = 0;
 
 	printl("cbuf_create\n");
 	if (unlikely(cbid < 0)) return 0;
 	CBUF_TAKE();
+
+	u64_t start, end;
+	rdtscll(start);
+
 	cci = cbuf_comp_info_get(spdid);
-	if (!cci) goto done;
+	if (unlikely(!cci)) goto done;
 
 	/* 
 	 * Client wants to allocate a new cbuf, but the meta might not
 	 * be mapped in.
 	 */
 	if (!cbid) {
-		struct cbuf_bin *bin;
+		/* TODO: check if have enough free memory: ask mem manager */
+		/*memory usage exceeds the target, block this thread*/
+		if (size + cci->allocated_size > cci->target_size) {
+			cbuf_shrink(cci, size);
+			if (size + cci->allocated_size > cci->target_size) {
+				cbuf_thread_block(cci, size);
+				return 0;
+			}
+		}
 
  		cbi = malloc(sizeof(struct cbuf_info));
-		if (!cbi) goto done;
 
-		/* Allocate and map in the cbuf. */
-		cbid = cmap_add(&cbufs, cbi);
+		if (unlikely(!cbi)) goto done;
+		/* Allocate and map in the cbuf. Discard inconsistent cbufs */
+		/* TODO: Find a better way to manage those inconsistent cbufs */
+		do {
+			cbid = cmap_add(&cbufs, cbi);
+			meta = cbuf_meta_lookup(cci, cbid);
```

Comment:
yes, we need cmap_del.
But we should record all inconsistent meta first, then del them together.
we cannot do cmap_del in the loop, as we will add to that slot again.